### PR TITLE
Concurrent Grading using asyncio

### PR DESCRIPTION
**Pull request by brhoades**
I've drafted concurrent grading using asyncio coroutines (threads instead of processes). By default, it's synchronous, but if -j is specified it uses that many worker threads.

It's currently NOT printing output as I'm not sure if that's desirable / if print locks stdout until it finishes.

Implements #21

Default (unchanged) behavior:
```
$ bin/grader grade hw1
INFO     Grading submissions for 6960054b3387
INFO:grade-it:Time to start the tests.
============================= test session starts ==============================
platform linux -- Python 3.4.3, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /tmp/tmp.nH26e9oICs, inifile:
collected 0 items

========================= no tests ran in 0.00 seconds =========================
INFO     Logging 6960054b3387 output as text
INFO     Wrote to grader/assignments/hw1/results/6960054b3387.13.log
INFO     Grading submissions for 9d7fe62204ad
INFO:grade-it:Time to start the tests.
============================= test session starts ==============================
platform linux -- Python 3.4.3, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /tmp/tmp.dMhPCDEGYY, inifile:
collected 0 items

========================= no tests ran in 0.00 seconds =========================
INFO     Logging 9d7fe62204ad output as text
INFO     Wrote to grader/assignments/hw1/results/9d7fe62204ad.13.log
INFO     Grading submissions for a65a6d7183c2
INFO:grade-it:Time to start the tests.
============================= test session starts ==============================
platform linux -- Python 3.4.3, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /tmp/tmp.myqboGzXZb, inifile:
collected 0 items

========================= no tests ran in 0.00 seconds =========================
INFO     Logging a65a6d7183c2 output as text
INFO     Wrote to grader/assignments/hw1/results/a65a6d7183c2.12.log
INFO     Grading submissions for dacb24920b25
INFO:grade-it:Time to start the tests.
============================= test session starts ==============================
platform linux -- Python 3.4.3, pytest-3.0.6, py-1.4.32, pluggy-0.4.0
rootdir: /tmp/tmp.s3cgnK0vx8, inifile:
collected 0 items

========================= no tests ran in 0.00 seconds =========================
INFO     Logging dacb24920b25 output as text
INFO     Wrote to grader/assignments/hw1/results/dacb24920b25.12.log
```

and with concurrency:
```
$ bin/grader grade -j2 hw1
INFO     Grading submission for 6960054b3387
INFO     Grading submission for 9d7fe62204ad
INFO     Logging 9d7fe62204ad output as text
INFO     Wrote to grader/assignments/hw1/results/9d7fe62204ad.14.log
INFO     Logging 6960054b3387 output as text
INFO     Wrote to grader/assignments/hw1/results/6960054b3387.14.log
INFO     Grading submission for a65a6d7183c2
INFO     Grading submission for dacb24920b25
INFO     Logging dacb24920b25 output as text
INFO     Wrote to grader/assignments/hw1/results/dacb24920b25.13.log
INFO     Logging a65a6d7183c2 output as text
INFO     Wrote to grader/assignments/hw1/results/a65a6d7183c2.13.log
```

### CODE REVIEW

**Comment by brhoades** [comment]
Output from threads is now, by default, printed unless suppress_output is enabled. Output is printed at the end of the thread by the thread (not as it is retrieved). This reduces the chance that two threads will print at once, but it's still possible.

Do we want to do a mutex lock on stdout here?

**Comment by LinuxMercedes** [comment]
I think it'd be nice to have printing synchronous, but not a requirement.

**Comment by brhoades** [comment]
@LinuxMercedes added print lock
@michaelwisely comments?

File: grader/grader/commands/grade/main.py [hunk 1]
```diff
@@ -13,8 +14,10 @@
 def setup_parser(parser):
     parser.add_argument('--rebuild', action='store_true',
                         help='Rebuild containers (if they exist).')
-    parser.add_argument('--suppress_output', action='store_false',
+    parser.add_argument('--suppress_output', action='store_true',
                         help='Don\'t display output.')
+    parser.add_argument('-j',  default="1",
```
Comment (michaelwisely) [inline grader/grader/commands/grade/main.py]: Can we add a long option (`--jobs`), too?

Comment (brhoades) [inline grader/grader/commands/grade/main.py]: Yep! Done.

**Comment by michaelwisely** [comment]
@brhoades i think there's a more idiomatic way to handle output without using a `Lock()`. Let me have a go at refactoring tonight.

File: grader/grader/commands/grade/async.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+
+def grade(submission, print_lock, rebuild, suppress_output):
+    """
+    Grade a single submission asynchronously. Locks to print
+    output if output is not suppressed.
+    """
+    logger.info("Grading submission for %s", submission.user_id)
+    output = submission.grade(
+        rebuild_container=rebuild, show_output=False)
```
Comment (michaelwisely) [inline grader/grader/commands/grade/async.py]: @brhoades is this an unbalanced parenthesis?

**Pull request closed without merging**